src/boot/efi/chid.c

```diff
@@ -0,0 +1,385 @@
+                                info->manufacturer,
+                                L"&",
+                                info->product_name,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 9:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                3,
+                                info->manufacturer,
+                                L"&",
+                                info->product_name);
+                break;
+        case 10:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                7,
+                                info->manufacturer,
+                                L"&",
+                                info->family,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 11:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                3,
+                                info->manufacturer,
+                                L"&",
+                                info->family);
+                break;
+        case 13:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                5,
+                                info->manufacturer,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 14:
+                status = hash_strings_sha1(
+                                &hash, (uint8_t *) &namespace, sizeof(namespace), 1, info->manufacturer);
+                break;
+        default:
+                return EFI_SUCCESS; /* Just keep empty to prevent match. */
+        }
+        if (EFI_STATUS_IS_ERROR(status))
+                return status;
+
+        memcpy(chid, hash, sizeof(*chid));
+
+        /* Convert the resulting CHID back to little-endian: */
+        chid->Data1 = __builtin_bswap32(chid->Data1);
+        chid->Data2 = __builtin_bswap16(chid->Data2);
+        chid->Data3 = __builtin_bswap16(chid->Data3);
```

Comment:
hmm, instead of using these builtins directly, i'd prefer we add some macros to util.h that create aliases for them that match the GNU macros we have in userspace, i.e. bswap_16(), bswap_32() and so on, see bswap(3) man page.